File: PrimeMinistersByJava/primeministers/Translator.java
package primeministers;

import javax.swing.JOptionPane;

import java.util.ArrayList;
import java.util.Calendar;

/**
 * トランスレータ：総理大臣のCSVファイルをHTMLページへと変換するプログラム。
 * 良好（2013年12月9日）
 */
public class Translator extends Object
{
	/**
	 * ここを作成してください。
	 * まず、次のページを参照しながら、スケルトン（スタブ）を作ることから始めましょう。
	 * http://www.cc.kyoto-su.ac.jp/~atsushi/Programs/CSV2HTML/PrimeMinistersJavaDoc/index.html
	 */
	
	/**
	 * CSVに由来するテーブルを記憶するフィールド
	 * 良好（2013年12月8日）
	 */
	private Table inputTable;
	
	/**
	 * HTMLに由来するテーブルを記憶するフィールド
	 * 良好（2013年12月8日）
	 */
	private Table outputTable;
	
	
	/**
	 * トランスレータのコンストラクタ。
	 * 良好（2013年12月8日）
	 */
	public Translator()
	{
		super();
		this.inputTable = new Table();
		this.outputTable = new Table();
		/*
		String days = this.computeNumberOfDays("1964年11月9日〜1972年7月7日");
		System.out.println("---- DUBUG = "+days+"---------");
		String days2 = this.computeNumberOfDays("2012年12月26日〜");
		System.out.println("---- DUBUG = "+days2+"---------");
		 */
		
		return;
	}
	
	/**
	 * 総理大臣のCSVファイルをHTMLページへ変換する。
	 * バグ（2013年12月8日）
	 */
	public void perform()
	{
		Downloader aDownload = new Downloader();
		aDownload.downloadCSV();
		this.inputTable = aDownload.table();
		
		aDownload.downloadImages();
		aDownload.downloadThumbnails();

		/* テーブルの確認用デバッガーもどき */
		//this.debugTable(); //後で消す
		
		Table aTable = this.table(this.inputTable);
		
		Writer aWriter = new Writer();
		aWriter.table(aTable);
		String aString = "総理大臣のCSVファイルからHTMLページへの変換を無事に完了しました。\n";
		JOptionPane.showMessageDialog(null, aString, "報告", JOptionPane.PLAIN_MESSAGE);
		
		return;
	}
	
	/**
	 * テープルの状態を確認するデバック用メソッド
	 * 後で消してね----------------------------!!
	 */
	private void debugTable()
	{
		for (String str : this.inputTable.attributes().keys())
		{
			System.out.print(str + " ");
		}
		
		for (Tuple aTuple : this.inputTable.tuples())
		{
			for (String aString : aTuple.values())
			{
				System.out.print(aString + " ");
			}
			System.out.println();
		}
	}
	
	/**
	 * 在位日数を計算して、それを文字列にして応答する。
	 * 良好（2013年12月9日）
	 */
	public String computeNumberOfDays(String periodString)
	{
		ArrayList<String> aStringArray = IO.splitString(periodString,"〜年月日");
		//From [0]年 [1]月 [2]日 : To [3]年 [4]月 [5]日
		
		int yearFrom, monthFrom, dayFrom;
		int yearTo, monthTo, dayTo;
		
		yearFrom = Integer.parseInt(aStringArray.get(0));
		monthFrom = Integer.parseInt(aStringArray.get(1));
		dayFrom = Integer.parseInt(aStringArray.get(2));
		
		if(aStringArray.size() < 4)
		{
			Calendar aCalendar = Calendar.getInstance();
			yearTo = aCalendar.get(Calendar.YEAR);        //現在の年を取得
			monthTo = aCalendar.get(Calendar.MONTH) + 1;  //現在の月を取得
			dayTo = aCalendar.get(Calendar.DATE) + 1;     //現在の日を取得
			
		}
		else
		{
			yearTo = Integer.parseInt(aStringArray.get(3));
			monthTo = Integer.parseInt(aStringArray.get(4));
			dayTo = Integer.parseInt(aStringArray.get(5));
		}
		
		Calendar aCalendarFrom = Calendar.getInstance();
		Calendar aCalendarTo = Calendar.getInstance();
		
		aCalendarFrom.set(yearFrom, monthFrom, dayFrom);
		aCalendarTo.set(yearTo, monthTo, dayTo);
		
		long from = aCalendarFrom.getTimeInMillis();
		long to = aCalendarTo.getTimeInMillis();
		long one_date_time = 1000*60*60*24;
		
		long diffDays = (to - from)/one_date_time;
		
		String days = String.format("%s",diffDays);
		
		return days;
	}
	
	/**
	 * サムネイル画像から画像へ飛ぶためのHTML文字列を生成して、それを応答する。
	 * 良好（2013年12月8日）
	 */
	public String computeStringOfImage(String aString,Tuple aTuple,int no)
	{
		ArrayList<String> values = aTuple.values();
		
		String aNo = values.get(aTuple.attributes().indexOfNo());
		String aImage = values.get(aTuple.attributes().indexOfImage());
		String aThumbnail = values.get(aTuple.attributes().indexOfThumbnail());
		
		String imageTag = "<a name="+aNo+" href="+aImage+"><img class=\"borderless\" src="+aThumbnail+" width=\"25\" height=\"32\" alt="+aNo+".jpg></a>";
		
		return imageTag;
	}
	/**
	 * 総理大臣のCSVファイルを基にしたテーブルから、HTMLページを基にするテーブルに変換して、
	 * それを応答する。
	 * 良好（2013年12月8日）
	 */
	public Table table(Table aTable)
	{

		Table htmlTable = new Table();
		htmlTable.attributes(new Attributes("output"));
		
		ArrayList<Tuple> tuples = aTable.tuples();
		
		for (Tuple aTuple : tuples)
		{
			ArrayList<String> values = aTuple.values();
			Attributes attribute = aTuple.attributes();
			ArrayList<String> output = new ArrayList<String>();
			
			output.add(values.get(attribute.indexOfNo()));//人目
			output.add(values.get(attribute.indexOfOrder()));//代
			output.add(values.get(attribute.indexOfName()));//氏名
			output.add(values.get(attribute.indexOfKana()));//ふりがな
			output.add(values.get(attribute.indexOfPeriod()));//在位期間
			
			output.add(this.computeNumberOfDays(values.get(attribute.indexOfPeriod())));//在位日数
			
			output.add(values.get(attribute.indexOfSchool()));//出身校
			output.add(values.get(attribute.indexOfParty()));//政党
			output.add(values.get(attribute.indexOfPlace()));//出身地
			output.add(this.computeStringOfImage(null, aTuple, Integer.valueOf(values.get(attribute.indexOfNo()))));//画像
			
			Tuple htmlTuple = new Tuple(htmlTable.attributes(), output);
			htmlTable.add(htmlTuple);
		}
		
		this.outputTable = htmlTable;
		
		return this.outputTable;
	}
	
}

File: PrimeMinistersByPython/primeministers/translator.py
#! /usr/bin/env python
# -*- coding: utf-8 -*-

import datetime
import os
import re

import table
import tuple

import csv

class Translator(object):
	"""トランスレータ：総理大臣のCSVファイルをHTMLページへと変換するプログラム。"""

	def __init__(self, input_table):
		"""トランスレータのコンストラクタ。"""
		self._input_table = input_table
		self._output_table = table.Table('output')
		return

	def compute_string_of_days(self, period):
		"""在位日数を計算して、それを文字列にして応答する。"""
		
		if period == '在位期間': return '在位日数'
		days = re.split(r'[〜年月日]*', period)
		days.pop()
		days = map(int, days)
		
		from_day = datetime.date(days[0], days[1], days[2])
		if len(days) > 4:
			to_day = datetime.date(days[3], days[4], days[5])
		else:
			d = datetime.datetime.today()
			to_day = datetime.date(d.year, d.month, d.day)
		diff = to_day - from_day
		
		return str(diff.days)

	def compute_string_of_image(self, tuple):
		"""サムネイル画像から画像へ飛ぶためのHTML文字列を作成して、それを応答する。"""
		
		keys = tuple.attributes().keys()
		values = tuple.values()
		a_no = values[keys.index('no')]
		a_image = values[keys.index('image')]
		a_thumbnail = values[keys.index('thumbnail')]
		imageTag = "<a name="+a_no+" href="+a_image+"><img class=\"borderless\" src="+a_thumbnail+" width=\"25\" height=\"32\" alt="+a_no+".jpg></a>"
		
		return imageTag

	def table(self):
		"""総理大臣のCSVファイルを基にしたテーブルから、HTMLページを基にするテーブルに変換して、それを応答する。"""
		html_table = table.Table('output')
		html_at = html_table.attributes()
		tuples = self._input_table.tuples()
		is_first = True
		
		for a_tuple in tuples:
			if is_first:
				is_first = False
			else:
				values = a_tuple.values()
				keys = a_tuple.attributes().keys()
				attributes = a_tuple.attributes()
				output = []
				output.append(values[keys.index('no')])
				output.append(values[keys.index('order')])
				output.append(values[keys.index('names')])
				output.append(values[keys.index('kana')])
				output.append(values[keys.index('period')])
				output.append(self.compute_string_of_days(values[keys.index('period')]))
				output.append(values[keys.index('school')])
				output.append(values[keys.index('party')])
				output.append(values[keys.index('birth')])
				output.append(self.compute_string_of_image(a_tuple))

				output_tuple = tuple.Tuple(html_at, output)
				self._output_table.add(output_tuple)
		
		return self._output_table
